### PR TITLE
Add 'insert_auto_ids' support to 'Batch'.

### PR DESCRIPTION
**Pull request by tseaver**
Follow-on to #509.

### CODE REVIEW

File: gcloud/datastore/batch.py [hunk 1]
```diff
@@ -159,7 +163,11 @@ def commit(self):
         however it can be called explicitly if you don't want to use a
         context manager.
         """
-        self.connection.commit(self._dataset_id, self.mutation)
+        response = self.connection.commit(self._dataset_id, self.mutation)
+        for new_key_pb, entity in zip(response.insert_auto_id_key,
+                                      self._auto_id_entities):
```
Comment (dhermes) [inline gcloud/datastore/batch.py]: @pcostell Are we gauranteed that a [`MutationResult`][1] will have the correct keys in the same order sent in the original `Mutation`? (i.e. Can the order change? Can there be partial success?)

[1]: https://github.com/GoogleCloudPlatform/google-cloud-datastore/blob/21e029b2657ebb15fef5429c90c4315326855fd1/proto/datastore_v1.proto#L409

**Comment by dhermes** [comment]
It looks fine but let's hear from Patrick before merging.

File: gcloud/datastore/batch.py (hunk 1 above)
Comment (pcostell) [inline gcloud/datastore/batch.py]: Yes the order will be the same. The commit will either all fail or all succeed, so no partial success.

Comment (dhermes) [inline gcloud/datastore/batch.py]: Yay! Thanks for the quick reply.

@tseaver Can you add that as a comment in the code?

**Comment by tseaver** [comment]
Second commit adds a method, `add_auto_id_entity`, to `Batch`, so that external callers (e.g., `Entity.save` or `datastore.put`) can handle the batch like a transaction.

File: gcloud/datastore/batch.py [hunk 2]
```diff
@@ -107,6 +108,25 @@ def mutation(self):
...
+        insert_auto_id entry in the mutation.  In order to make sure we
+        update the Entity once the transaction is committed, we need to
+        keep track of which entities to update (and the order is
+        important).
+
+        When you call ``save()`` on an entity inside a transaction, if
+        the entity has a partial key, it adds itself to the list of
+        entities to be updated once the transaction is committed by
+        calling this method.
+        """
```
Comment (dhermes) [inline gcloud/datastore/batch.py]: Needs `:param:` and `:raises:`

(Sorry commented on the commit not the PR)

**Comment by tseaver** [comment]
I will wait for Patrick's reply.

**Comment by dhermes** [comment]
@tseaver He already replied, see my [comment][1] above.

[1]: https://github.com/GoogleCloudPlatform/gcloud-python/pull/515#discussion_r22670476

File: gcloud/datastore/batch.py (hunk 1 above)
Comment (tseaver) [inline gcloud/datastore/batch.py]: 5c10b105de35235860a0db5af0c59c6a8ec95256

**Comment by dhermes** [comment]
LGTM.

The behavior of the `Commit` RPC is a relief :)

**Comment by tseaver** [comment]
Yup.  I guess the difference is not "partial success" but rather "eventual consistency" (no guarantees that the change will be visible immediately).

**Pull request merged**